### DISCUSSION TITLE
How to handle type of Element.Properties?

### DISCUSSION THREAD

**Discussion by plastic041** [rehypejs/rehype#72] [Q&A] [answered]
`Element`s parsed by `rehype-parse` has `Properties`, which is typed as `Properties | undefined`.
However, `rehype-parse` parses some elements' properties as `{}`, even if there is no properties at all.

example:

```javascript
const file = `<img>`;

const logNodes = () => {
  return (tree) => {
    visit(tree, "element", (node) => {
      console.log(`
        ${node.tagName}: ${JSON.stringify(node.properties)}
      `); // should be "img: undefined", but actually logs "img: {}"

      if (node.properties) { //  unnecessary because node.properties is never undefined, but without check it gives type error.
        // do something...
      }
    });
  };
};

unified()
  .use(rehypeParse, { fragment: true })
  .use(logNodes)
  .use(rehypeStringify)
  .process(file);
```

The problem is, I cannot utilize `Properties` without checking its type, even if I am sure that it is never `undefined`.
I am currently using `@ts-ignore` to supress this issue.
Is this an expected behavior? If it is, how can I handle this without using `@ts-ignore`?

Codepen link: https://codepen.io/plastic041/pen/powYBXQ

**Comment by wooorm** [comment, marked as answer]
If you type `tree` as a `hast.Root`, everything becomes typed.

`properties` should always be `Properties`, not `undefined`. However, people sometimes forget to add it in their plugins. That’s why it’s recommended to use an if check to make sure it exists

**Reply by plastic041** [reply]
![image](https://user-images.githubusercontent.com/17811025/135586446-e24c658a-86c9-49cc-be60-160104a17f63.png)

It still says `Element.properties` is `Properties | undefined`, even if I typed `tree` as `hast.Root`. What am I missing?

**Reply by wooorm** [reply]
My second paragraph?

**Reply by plastic041** [reply]
But you said `properties` should always be `Properties`? It is still `Properties | undefined`.

**Reply by wooorm** [reply]
Indeed, I said *should*. But people (who don’t use TypeScript) forget. That’s why it’s typed as optional: to reflect the real world. So that in your code you can check for whether it exists or not and prevent runtime errors.

**Reply by plastic041** [reply]
I see, I misinterpreted your reply. Thank you for the answer!

**Reply by wooorm** [reply]
No problem! All the best and good luck.

P.S. we have some guides on unified with typescript: https://unifiedjs.com/learn/guide/syntax-trees-typescript/ and https://unifiedjs.com/learn/recipe/tree-traversal-typescript/. Might be useful?

**Reply by plastic041** [reply]
Didn't know that. Thanks, I'll check 😀
